pandas/_libs/internals.pyx

```diff
@@ -793,12 +796,14 @@ cdef class BlockManager:
             ndarray blknos, blklocs
 
         nbs = []
+        nrefs = []
         for blk in self.blocks:
             nb = blk.getitem_block_index(slobj)
             nbs.append(nb)
+            nrefs.append(weakref.ref(blk))
```

Comment:
Good question. I don't remember if there was a specific technical reason to do so, but I think it seemed the easier option (since this is all handled at the BlockManager level). When keeping those references / when checking for references, it would otherwise be an additional level of indirection to check the blk.values instead of blk itself.

I suppose this would generally be the same, but indeed except in places where `blk.values` is re-set in place. I actually added one such case in this PR  in `Block.set_values` (which is called in `BlockManager.iset`). In that case I should probably rather copy the Block and replace the block in the BM with the copied block, instead of copying the values in the existing block. 
Do you know of other places where we currently we re-set blk.values?